proxy/stats_handler.go

```diff
@@ -0,0 +1,215 @@
+package proxy
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/gorilla/websocket"
+
+	"github.com/rancherio/websocket-proxy/common"
+)
+
+type StatsHandler struct {
+	backend         backendProxy
+	parsedPublicKey interface{}
+}
+
+func (h *StatsHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	multiHost := false
+
+	if strings.HasSuffix(req.URL.Path, "project/") || strings.HasSuffix(req.URL.Path, "service/") {
+		multiHost = true
+	}
+
+	hostKeys, urls, authed := h.auth(req, multiHost)
+	if !authed {
+		return
+	}
+
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+	ws, err := upgrader.Upgrade(rw, req, nil)
+	if err != nil {
+		http.Error(rw, "Failed to upgrade connection.", 500)
+		return
+	}
+
+	msgKeys := make([]string, len(hostKeys))
+	respChannels := make([]<-chan common.Message, len(hostKeys))
+
+	var mutex sync.Mutex
+	var countMutex sync.Mutex
+
+	doneCounter := len(hostKeys)
+
+	for i, hostKey := range hostKeys {
+		var err error
+		msgKeys[i], respChannels[i], err = h.backend.initializeClient(hostKey)
+		if err != nil {
+			h.closeConnection(ws)
+			return
+		}
+
+		// Send response messages to client
+		go func(pos int) {
+			for {
+				message, ok := <-respChannels[pos]
+				if !ok {
+					return
+				}
+				switch message.Type {
+				case common.Body:
+					ws.SetWriteDeadline(time.Now().Add(10 * time.Second))
+					mutex.Lock()
+					if err := ws.WriteMessage(1, []byte(message.Body)); err != nil {
+						for j, hostKey := range hostKeys {
+							h.backend.closeConnection(hostKey, msgKeys[j])
+						}
+						h.closeConnection(ws)
+					}
+					mutex.Unlock()
+				case common.Close:
+					countMutex.Lock()
+					doneCounter--
+					if doneCounter == 0 {
+						h.closeConnection(ws)
+					}
+					h.backend.closeConnection(hostKeys[pos], msgKeys[pos])
```

Comment:
What if the respChannel is closed without a common.Close message ever being sent? The doneCounter would not get decremented.

I think you should put the doneCounter and closeConnection logic in a deferred function inside this go routine, so that if the go routine exits for any reason, the counter gets decremented properly.